ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -537,9 +537,14 @@ let retrieve_wlb_recommendations ~__context ~vm ~snapshot =
   List.iter
     (fun (h, r) ->
        try
-         assert_can_boot_here ~__context ~self:vm ~host:h ~snapshot ~do_cpuid_check:false ();
+         assert_can_boot_here ~__context ~self:vm ~host:h ~snapshot ~do_cpuid_check:true ();
          Hashtbl.replace recs h r;
        with
+       | Api_errors.Server_error(x, _) when x=Api_errors.vm_incompatible_with_this_host ->
+         try
+           let zeroScoreReasonVmIncompatible = "2024" in
+           Hashtbl.replace recs h [List.nth r 0; "0"; List.nth r 2; zeroScoreReasonVmIncompatible]
+         with Invalid_argument _ ->  debug "Missing element of recommendation of host %s!" (Ref.string_of h)
```

Comment:
Instead of try/with you could also use a match:
```ocaml
| Api_errors.Server_error(x, y) when x=Api_errors.vm_incompatible_with_this_host ->
   begin match r with
   |  id :: _ :: rec_id :: _ -> 
      let zeroScoreReasonVmIncompatible = "2024" in
      Hashtbl.replace recs h [id; "0.0"; rec_id; zeroScoreReasonVmIncompatible]
   | _ ->
    debug "Missing element of recommendation of host %s!" (Ref.string_of h);
    Hashtbl.replace recs h (x :: y)
   end
```

Also replace `0` with `0.0`, that is what some of the rest of the code seems to expect, see in function `choose_host_for_vm` this line: `"WLB"; "0.0"; rec_id; zero_reason]`

Reply:
"r" can be a 3 or 4 elements list, see function retrieve_vm_recommendations.
So I use list element access.
OK, I will try match.

I set it to "0.0" first, but WLB set zero score to "0".
OK. I will set it to 0.0 to match the logic in xapi.